src/tod_manager.cpp

```diff
@@ -536,3 +548,29 @@ bool tod_manager::is_time_left() const
 {
 	return num_turns_ == -1 || turn_ <= num_turns_;
 }
+
+int tod_manager::calculate_best_liminal_bonus(const std::vector<time_of_day>& schedule) const
+{
+	int fearless_chaotic = 0;
+	int fearless_lawful = 0;
```

Comment:
I'm kinda confused about the purpose of these. Don't fearless units ignore ToD bonuses and maluses?

Reply:
Fearless only ignores maluses, not bonuses. (That would be neutral.)

Reply:
Okay, so... `fearless_chaotic` is the sum of all bonuses for a chaotic unit in the schedule, and `fearless_lawful` is the same byt for lawful units. And `bonuses` contains the abs of all the ToD modifiers in the schedule.

With the built-in schedules, these end up being:

* **Default**: fearless_chaotic = 50, fearless_lawful = 50
* **Summer**: fearless_chaotic = 25, fearless_lawful = 75
* **Winter**: fearless_chaotic = 75, fearless_lawful = 25
* **Indoors**: fearless_chaotic = 0, fearless_lawful = 0
* **Underground**: fearless_chaotic = 25, fearless_lawful = 0
* **Deep underground**: fearless_chaotic = 30, fearless_lawful = 0
* **24-hour**: fearless_chaotic = 215, fearless_lawful = 215
* **Two-suns**: fearless_chaotic = 75, fearless_lawful = 150

Okay, and then for each of the bonuses... you calculate `liminal_effect` which is similar to `fearless_chaotic` and `fearless_lawful` but this time for non-fearless liminal units. (Come to think of it, fearless actually has no effect on this new type of liminal?) Instead of the actual max liminal bonus (which is what this function is calculating), you pass the abs-bonus for this time slot as the max liminal bonus, which means what you get back is `max(0, bonus - abs(lawful_bonus))`, but `bonus` is itself `abs(lawful_bonus)`, so doesn't this always return 0? Am I missing something here?

Anyway, assuming I'm right, that means you then check if `abs(target) < delta`, which will never pass because they start equal, which would always give you a final result of 0. If it _did_ pass, what you'd be doing is finding the minimum of ..._something_ that I don't quite understand, and the final result ends up being the absolute value of one of the existing time's lawful bonuses.

So unless I _am_ missing something, I think the algorithm needs a lot more work. Can you at least explain what you were _intending_ to achieve in some detail?

Taking the minimum abs-bonus of the time schedule doesn't seem like it'll work (because that would give you 5 in the 24-hour schedule); on the other hand I'm not sure the maximum abs-bonus (which would at least give you 25 in all built-in schedules, except maybe deep underground) is the best route either.

I'm also not so sure that basing it on the _entire_ schedule is the best course of action, though the only other option I can think of is basing it on the two adjacent time slots which isn't any better (it'd probably yield 5 or similar in the 24-hour schedule).

Reply:
You have the fearless_chaotic and fearless_lawful right. 

Then you once again make the assumption that tod_manager is a time slot, when in fact it is a scenario-wide object that manages the whole schedule (and the "current time" index, not used here).

I would like a {0, -20, -20, 0, 20, 20} schedule to give a liminal effect = 20 - abs(lawful_bonus).
The constant in the formula is a scenario-wide info. It may be changed at the scenario level. But if it is left unchanged, I assume it means that the scenario designer wishes Default+Dunefolk to be as balanced as possible, so I compute a constant that seems to fit.

Reply:
Okay, so... for each absolute lawful bonus in the schedule, you go through the entire schedule _again_ and add up the total of what the liminal bonus would be if the bonus under consideration were the maximum liminal bonus.

Or to put it another way... for each possible absolute lawful bonus in the schedule, you take that as the max liminal bonus, and then run through the whole schedule _again_, adding up the total liminal effect across all times.

Anyway, however you want to describe it, this gives you another sum-of-times value similar to the fearless values, which are used here as a seed of sorts. I can't think of how to describe this in the general case, but taking the default schedule as an example, you'd have to loop twice for the values 0 and 25. On the first loop, `target` is 25, and `liminal_effect` would be, uhh... 50? So on that loop, `result` is not updated because `abs(25 - 50) == delta`. On the second loop, `liminal_effect` would be ... 150? I have a feeling I might've messed up my math here, because that doesn't make sense - `abs(25 - 150) > delta`, so `result` would be 0.

I feel like I'm sort of starting to understand, but I guess I'm still confused...

Reply:
Note that liminal_effect is reset to 0 on each iteration over bonuses.

Let me detail some examples, I hope you'll understand the process.
For default schedule ({0, -25, -25, 0, 25, 25}): 
target = max(fearless_chaotic, fearless_awful) = 50.
bonuses = {0, 25}
with 0, liminal [fearless] effects over schedule are all 0, sum 0, so not selected.
with 25, liminal effects over schedule are {25, 0, 0, 25, 0, 0}, sum 50, selected (perfect match, delta becomes 0).

Imagine a long twilight schedule {0, 0, 5, 15, 20, 25, 20, 15, 5, 0, 0, -5, -15, -20, -25, -20, -15, -5} 
target = 105.
bonuses = {0, 5, 15, 20, 25}
with 0, liminal_effect = 0
with 5, liminal_effect = 20 better
with 15, liminal_effect = 100 even better, delta becomes 5.
with 20, liminal_effect = 140 is too strong, 
with 25, liminal_effect = 240 is far too strong.

For me, assuming an era containing liminal units is balanced for standard schedule means assuming the power of liminal should stay close to the power of always_fearless. So, in an imaginary long twilight schedule, it should not become "era of all mighty liminals", and the calculation tones down liminal, by selecting 15 as the maximum bonus it gives.

Reply:
Hmm, I see. So the goal is that the _total_ liminal bonus over all time slots is no greater than the max bonus that a chaotic or lawful unit can get? It's a little weird that liminal doesn't get a fixed bonus, but I suppose the actual bonus would be shown somewhere in the UI anyway.

It's also interesting that the chosen bonus for liminal is always the absolute value of a bonus that exists in the schedule for either lawful or chaotic. I think this algorithm can come up with a sane fallback default, but if you're using modifiers other than ±25 I think the fallback would tend not to be the most logical (that is, you might want liminal to rise to a peak in a manner similar to how lawful ans chaotic do in your example long twilight schedule).